File: app/local_plugins/agent_sts_chat/android/src/main/kotlin/com/aiagent/agent_sts_chat/StsChatAgentSession.kt
package com.aiagent.agent_sts_chat

import android.content.Context
import android.util.Log
import com.aiagent.local_db.AppDatabase
import com.aiagent.local_db.entity.MessageEntity
import com.aiagent.plugin_interface.*
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.MutableStateFlow
import java.util.UUID

/**
 * StsChatAgentSession — STS (端到端语音对话) Agent 原生实现
 *
 * 使用 NativeStsService 通过 WebSocket 进行实时语音对话。
 * 服务端完成 ASR → LLM → TTS 全流程，客户端只负责音频收发。
 *
 * 简化状态机：IDLE → CONNECTED → IDLE
 *
 * 生命周期：
 *   initialize → connect(WebSocket) → setInputMode("call") → startAudio
 *   → [实时双向音频流] → setInputMode("text") → stopAudio → release
 *
 * 从 AgentSession.kt 的 isStsAgent 分支提取。
 */
class StsChatAgentSession : NativeAgent {

    companion object {
        private const val TAG = "StsChatAgentSession"
    }

    override val agentType = "sts"

    private enum class State { IDLE, CONNECTED, ERROR }

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private val _state = MutableStateFlow(State.IDLE)

    private lateinit var stsService: NativeStsService
    private lateinit var db: AppDatabase
    private lateinit var eventSink: AgentEventSink
    private lateinit var config: NativeAgentConfig
    private lateinit var appContext: Context

    private var connectJob: Job? = null
    private var inputMode: String = "text"

    /** 当前 AI 回复的消息 ID（同一轮回复累加到同一条消息） */
    @Volatile private var currentAssistantId: String? = null
    /** 暂存部分识别的用户文本 */
    @Volatile private var pendingUserText: String = ""
    /** 上次已向 UI 发送的累积文本长度，用于计算 delta */
    @Volatile private var lastSentLength: Int = 0

    // ─────────────────────────────────────────────────
    // NativeAgent 接口实现
    // ─────────────────────────────────────────────────

    override fun initialize(config: NativeAgentConfig, eventSink: AgentEventSink, context: Context) {
        this.config = config
        this.eventSink = eventSink
        this.inputMode = config.inputMode
        this.appContext = context.applicationContext
        this.db = AppDatabase.getInstance(context)

        // Create STS service from NativeServiceRegistry
        stsService = NativeServiceRegistry.createSts(config.stsVendor ?: "volcengine")

        // Initialize STS service with config
        stsService.initialize(config.stsConfigJson ?: "{}", context)

        Log.d(TAG, "initialized: agentId=${config.agentId} stsVendor=${config.stsVendor}")
    }

    override fun connectService() {
        Log.d(TAG, "connectService: agentId=${config.agentId}")
        connectJob?.cancel()

        // 重建 STS service（上一次 disconnectService 调用了 release）
        stsService = NativeServiceRegistry.createSts(config.stsVendor ?: "volcengine")
        stsService.initialize(config.stsConfigJson ?: "{}", appContext)

        connectJob = scope.launch {
            try {
                stsService.connect(stsCallback)
            } catch (e: CancellationException) {
                // normal cancellation
            } catch (e: Exception) {
                Log.e(TAG, "STS connect failed: ${e.message}")
                transitionTo(State.ERROR)
                eventSink.onError(config.agentId, "sts_connect_error", e.message ?: "Unknown error", null)
                eventSink.onAgentReady(
                    config.agentId, ready = false,
                    errorCode = "sts_connect_error",
                    errorMessage = e.message ?: "Unknown error",
                )
            }
        }
    }

    override fun disconnectService() {
        Log.d(TAG, "disconnectService: agentId=${config.agentId}")
        connectJob?.cancel()
        stsService.release()
        transitionTo(State.IDLE)
        eventSink.onConnectionStateChanged(config.agentId, "disconnected")
    }

    override fun sendText(requestId: String, text: String) {
        // STS 是纯语音模式，不支持文本输入
        Log.w(TAG, "sendText called but STS is voice-only, ignoring: $text")
    }

    override fun startListening() {
        // STS 模式不使用独立的 STT 监听，音频通过 WebSocket 直接发送
        Log.d(TAG, "startListening: delegating to startAudio")
        stsService.startAudio()
    }

    override fun stopListening() {
        Log.d(TAG, "stopListening: delegating to stopAudio")
        stsService.stopAudio()
    }

    override fun setInputMode(mode: String) {
        Log.d(TAG, "setInputMode: $mode")
        inputMode = mode
        when (mode) {
            "call" -> {
                // 开始发送麦克风音频（需先调用 connectService 建立 WebSocket）
                stsService.startAudio()
            }
            "short_voice" -> { /* 按住说话，由 UI 调用 startListening/stopListening */ }
            else -> {
                // 停止发送麦克风音频，WebSocket 保持连接
                stsService.stopAudio()
            }
        }
    }

    override fun interrupt() {
        // STS 模式：仅清空 TTS 播放缓冲，WebSocket 保持连接
        stsService.interrupt()
        transitionTo(State.CONNECTED)
    }

    override fun release() {
        connectJob?.cancel()
        scope.cancel()
        stsService.release()
    }

    // ─────────────────────────────────────────────────
    // 外部音频源（AI 助理 / 通话翻译等场景）—— 透传到底层 NativeStsService
    //
    // 与 setInputMode("call") 触发的 self-mic 路径互斥；编排器在 AI 助理场景
    // 显式以 inputMode='external' 初始化，并在 connect 完成后调
    // startExternalAudio 接管音源，绝不能再走 self-mic（会抢 RECORD_AUDIO 与 SCO）。
    // ─────────────────────────────────────────────────

    override fun externalAudioCapability(): ExternalAudioCapability =
        stsService.externalAudioCapability()

    override fun startExternalAudio(format: ExternalAudioFormat, sink: ExternalAudioSink) {
        stsService.startExternalAudio(format, sink)
    }

    override fun pushExternalAudioFrame(frame: ByteArray) {
        stsService.pushExternalAudioFrame(frame)
    }

    override fun stopExternalAudio() {
        stsService.stopExternalAudio()
    }

    // ─────────────────────────────────────────────────
    // STS 回调 → AgentEventSink 事件转换
    // ─────────────────────────────────────────────────

    private val stsCallback = object : StsCallback {
        override fun onConnected() {
            transitionTo(State.CONNECTED)
            eventSink.onConnectionStateChanged(config.agentId, "connected")
            eventSink.onAgentReady(config.agentId, ready = true)
            Log.d(TAG, "WebSocket connected, inputMode=$inputMode")
            // 如果用户已切换到 call 模式但连接还没好，现在自动启动音频
            if (inputMode == "call") {
                stsService.startAudio()
            }
        }

        override fun onSttPartialResult(text: String) {
            // 部分识别 → 覆盖暂存，发 partialResult 供 UI 实时预览
            pendingUserText = text
            eventSink.onSttEvent(SttEventData(
                config.agentId, requestId = "", kind = "partialResult", text = text))
        }

        override fun onSttFinalResult(text: String) {
            // 最终识别 → 定稿用户消息，重置 AI 回复 ID（下次 sentenceDone 开新消息）
            pendingUserText = ""
            currentAssistantId = null

            val reqId = UUID.randomUUID().toString()
            eventSink.onSttEvent(SttEventData(
                config.agentId, requestId = reqId, kind = "finalResult", text = text))

            scope.launch {
                runCatching {
                    val now = System.currentTimeMillis()
                    db.messageDao().insert(MessageEntity(
                        id = reqId, agentId = config.agentId,
                        role = "user", content = text, status = "done",
                        createdAt = now, updatedAt = now,
                    ))
                }
            }
        }

        override fun onTtsAudioChunk(pcmData: ByteArray) {
            // TTS 音频由 STS 服务内部播放
        }

        override fun onChatPartialResult(cumulativeText: String) {
            // 流式 token 到来：计算增量并推送到 UI
            val delta = cumulativeText.substring(lastSentLength)
            if (delta.isEmpty()) return

            val isNew = currentAssistantId == null
            val msgId = currentAssistantId ?: UUID.randomUUID().toString().also { currentAssistantId = it }

            if (isNew) {
                // 首个 token：在 DB 插入 streaming 消息
                scope.launch {
                    runCatching {
                        val now = System.currentTimeMillis()
                        db.messageDao().insert(MessageEntity(
                            id = msgId, agentId = config.agentId,
                            role = "assistant", content = cumulativeText, status = "streaming",
                            createdAt = now, updatedAt = now,
                        ))
                    }
                }
            } else {
                // 后续 token：追加内容到 DB
                scope.launch {
                    runCatching {
                        db.messageDao().appendContent(msgId, delta, System.currentTimeMillis())
                    }
                }
            }

            lastSentLength = cumulativeText.length
            eventSink.onLlmEvent(LlmEventData(
                config.agentId, requestId = msgId, kind = "firstToken", textDelta = delta))
        }

        override fun onSentenceDone(text: String) {
            // AI 回复完成：若没有经过流式路径则兜底发完整文本；最终发 done 关闭气泡
            val msgId = currentAssistantId
            lastSentLength = 0

            if (msgId == null) {
                // 无流式更新（兜底）：整句一次性发出
                val id = UUID.randomUUID().toString()
                currentAssistantId = id
                eventSink.onLlmEvent(LlmEventData(
                    config.agentId, requestId = id, kind = "firstToken", textDelta = text))
                scope.launch {
                    runCatching {
                        val now = System.currentTimeMillis()
                        db.messageDao().insert(MessageEntity(
                            id = id, agentId = config.agentId,
                            role = "assistant", content = text, status = "done",
                            createdAt = now, updatedAt = now,
                        ))
                    }
                }
                eventSink.onLlmEvent(LlmEventData(
                    config.agentId, requestId = id, kind = "done", fullText = text))
                currentAssistantId = null
            } else {
                // 已有流式气泡：直接 done
                currentAssistantId = null
                eventSink.onLlmEvent(LlmEventData(
                    config.agentId, requestId = msgId, kind = "done", fullText = text))
                scope.launch {
                    runCatching {
                        db.messageDao().updateStatus(msgId, "done", System.currentTimeMillis())
                    }
                }
            }
        }

        override fun onDisconnected() {
            transitionTo(State.IDLE)
            eventSink.onConnectionStateChanged(config.agentId, "disconnected")
            Log.d(TAG, "WebSocket disconnected")
        }

        override fun onError(code: String, message: String) {
            transitionTo(State.ERROR)
            eventSink.onConnectionStateChanged(config.agentId, "error", message)
            eventSink.onError(config.agentId, code, message, null)
            eventSink.onAgentReady(config.agentId, ready = false, errorCode = code, errorMessage = message)
        }

        override fun onSpeechStart() {
            // 用户开口 → 当前 AI 回复被打断，标记 done，重置 ID
            val prevId = currentAssistantId
            lastSentLength = 0
            if (prevId != null) {
                eventSink.onLlmEvent(LlmEventData(
                    config.agentId, requestId = prevId, kind = "done"))
                scope.launch { runCatching { db.messageDao().updateStatus(prevId, "done", System.currentTimeMillis()) } }
                currentAssistantId = null
            }
            eventSink.onSttEvent(SttEventData(
                config.agentId, requestId = "", kind = "vadSpeechStart"))
        }

        override fun onStateChanged(state: String) {
            eventSink.onStateChanged(config.agentId, state, null)
        }
    }

    // ─────────────────────────────────────────────────
    // 内部方法
    // ─────────────────────────────────────────────────

    private fun transitionTo(newState: State) {
        _state.value = newState
        eventSink.onStateChanged(
            config.agentId,
            newState.name.lowercase(),
            null,
        )
    }
}

File: apps/client/local_plugins/agents/agent_sts_chat/ios/Classes/StsChatAgentSession.swift
import Foundation
import ai_plugin_interface
import os.log

/// STS (end-to-end speech-to-speech) agent session — iOS port of the Kotlin
/// `StsChatAgentSession`. Persistence (message DB) stays on the Dart side, so
/// this class is a thin event-translation layer over `NativeStsService`.
///
/// Lifecycle:
///   initialize → connectService (open WebSocket) → setInputMode("call")
///   → [bidirectional audio] → setInputMode("text") → disconnectService → release
public final class StsChatAgentSession: NativeAgent {

    public let agentType = "sts-chat"

    private static let log = OSLog(subsystem: "com.aiagent.agent_sts_chat", category: "Session")

    fileprivate enum State: String { case idle, connected, error }

    private let stateLock = NSLock()
    private var state: State = .idle

    private var config: NativeAgentConfig!
    private weak var eventSink: AgentEventSink?
    private var inputMode: String = "text"

    private var stsService: NativeStsService?

    /// Current assistant-message id; same id is reused across stream tokens
    /// of one reply, then cleared on `onSentenceDone` / `onSpeechStart`.
    private var currentAssistantId: String?
    /// Length of the cumulative chat text already pushed to UI — used to
    /// derive `firstToken.textDelta` from vendor snapshots.
    private var lastSentLength: Int = 0

    public init() {}

    // ── NativeAgent lifecycle ─────────────────────────────────────────────

    public func initialize(config: NativeAgentConfig, eventSink: AgentEventSink) {
        self.config = config
        self.eventSink = eventSink
        self.inputMode = config.inputMode

        do {
            let svc = try NativeServiceRegistry.shared.createSts(config.stsVendor ?? "volcengine")
            svc.initialize(configJson: config.stsConfigJson ?? "{}")
            self.stsService = svc
            os_log("initialized agentId=%{public}@ stsVendor=%{public}@ inputMode=%{public}@",
                   log: Self.log, type: .debug,
                   config.agentId, config.stsVendor ?? "volcengine", config.inputMode)
        } catch {
            os_log("createSts failed: %{public}@",
                   log: Self.log, type: .error, error.localizedDescription)
            eventSink.onError(
                sessionId: config.agentId,
                errorCode: "sts_init_failed",
                message: error.localizedDescription,
                requestId: nil
            )
        }
    }

    public func connectService() {
        os_log("connectService: agentId=%{public}@", log: Self.log, type: .debug, config.agentId)

        // Recreate STS service in case the previous disconnectService released it.
        do {
            let svc = try NativeServiceRegistry.shared.createSts(config.stsVendor ?? "volcengine")
            svc.initialize(configJson: config.stsConfigJson ?? "{}")
            self.stsService = svc
        } catch {
            os_log("connectService createSts failed: %{public}@",
                   log: Self.log, type: .error, error.localizedDescription)
            transitionTo(.error)
            eventSink?.onError(
                sessionId: config.agentId,
                errorCode: "sts_init_failed",
                message: error.localizedDescription,
                requestId: nil
            )
            eventSink?.onAgentReady(
                sessionId: config.agentId,
                ready: false,
                errorCode: "sts_init_failed",
                errorMessage: error.localizedDescription
            )
            return
        }

        stsService?.connect(callback: StsCallbackAdapter(owner: self))
    }

    public func disconnectService() {
        os_log("disconnectService: agentId=%{public}@", log: Self.log, type: .debug, config.agentId)
        stsService?.release()
        transitionTo(.idle)
        eventSink?.onConnectionStateChanged(
            sessionId: config.agentId,
            state: "disconnected",
            errorMessage: nil
        )
    }

    public func sendText(requestId: String, text: String) {
        os_log("sendText ignored — STS is voice-only (text=%{public}@)",
               log: Self.log, type: .info, String(text.prefix(60)))
    }

    public func startListening() {
        os_log("startListening → stsService.startAudio", log: Self.log, type: .debug)
        stsService?.startAudio()
    }

    public func stopListening() {
        os_log("stopListening → stsService.stopAudio", log: Self.log, type: .debug)
        stsService?.stopAudio()
    }

    public func setInputMode(_ mode: String) {
        os_log("setInputMode: %{public}@", log: Self.log, type: .debug, mode)
        inputMode = mode
        switch mode {
        case "call":
            stsService?.startAudio()
        case "short_voice":
            break  // UI drives startListening/stopListening
        default:
            stsService?.stopAudio()
        }
    }

    public func interrupt() {
        stsService?.interrupt()
        transitionTo(.connected)
    }

    public func release() {
        stsService?.release()
        stsService = nil
    }

    // ── External audio source (call translation / face-to-face) ───────────

    public func externalAudioCapability() -> ExternalAudioCapability {
        stsService?.externalAudioCapability() ?? .unsupported
    }

    public func startExternalAudio(format: ExternalAudioFormat, sink: ExternalAudioSink) throws {
        guard let svc = stsService else {
            throw NativeServiceError.invalidConfig("sts service not initialised")
        }
        try svc.startExternalAudio(format: format, sink: sink)
    }

    public func pushExternalAudioFrame(_ frame: Data) {
        stsService?.pushExternalAudioFrame(frame)
    }

    public func stopExternalAudio() {
        stsService?.stopExternalAudio()
    }

    // ── Internal helpers ──────────────────────────────────────────────────

    fileprivate func transitionTo(_ newState: State) {
        stateLock.lock()
        state = newState
        stateLock.unlock()
        eventSink?.onStateChanged(
            sessionId: config.agentId,
            state: newState.rawValue,
            requestId: nil
        )
    }

    fileprivate func sink() -> AgentEventSink? { eventSink }
    fileprivate func agentId() -> String { config.agentId }
    fileprivate func currentInputMode() -> String { inputMode }
    fileprivate func currentService() -> NativeStsService? { stsService }

    // Bridges callback-only mutations from the adapter back into the session.
    fileprivate func setCurrentAssistantId(_ id: String?) { currentAssistantId = id }
    fileprivate func getCurrentAssistantId() -> String? { currentAssistantId }
    fileprivate func setLastSentLength(_ n: Int) { lastSentLength = n }
    fileprivate func getLastSentLength() -> Int { lastSentLength }
}

// MARK: - StsCallback adapter

/// Translates `StsCallback` (vendor-side) into `AgentEventSink` events.
/// Kept as a separate class so the session can hand a strong reference to the
/// vendor service without creating a retain cycle.
private final class StsCallbackAdapter: StsCallback {
    private weak var owner: StsChatAgentSession?

    init(owner: StsChatAgentSession) { self.owner = owner }

    func onConnected() {
        guard let owner = owner else { return }
        owner.transitionTo(.connected)
        owner.sink()?.onConnectionStateChanged(
            sessionId: owner.agentId(),
            state: "connected",
            errorMessage: nil
        )
        owner.sink()?.onAgentReady(sessionId: owner.agentId(), ready: true)
        // If UI already requested call mode before connect completed, start now.
        if owner.currentInputMode() == "call" {
            owner.currentService()?.startAudio()
        }
    }

    func onSttPartialResult(text: String) {
        guard let owner = owner else { return }
        owner.sink()?.onSttEvent(SttEventData(
            sessionId: owner.agentId(),
            requestId: "",
            kind: "partialResult",
            text: text
        ))
    }

    func onSttFinalResult(text: String) {
        guard let owner = owner else { return }
        owner.setCurrentAssistantId(nil)
        let reqId = UUID().uuidString
        owner.sink()?.onSttEvent(SttEventData(
            sessionId: owner.agentId(),
            requestId: reqId,
            kind: "finalResult",
            text: text
        ))
    }

    func onTtsAudioChunk(pcmData: Data) {
        // TTS audio is played by the vendor service; no agent-level fan-out.
    }

    func onChatPartialResult(cumulativeText: String) {
        guard let owner = owner else { return }
        var last = owner.getLastSentLength()
        // Defensive: some vendors rewrite the snapshot shorter mid-stream.
        if last > cumulativeText.count { last = 0 }
        let startIdx = cumulativeText.index(cumulativeText.startIndex, offsetBy: last)
        let delta = String(cumulativeText[startIdx...])
        if delta.isEmpty { return }

        let isNew = owner.getCurrentAssistantId() == nil
        let msgId: String
        if let existing = owner.getCurrentAssistantId() {
            msgId = existing
        } else {
            msgId = UUID().uuidString
            owner.setCurrentAssistantId(msgId)
        }
        _ = isNew  // parity with Android (Dart side derives "first chunk" from requestId state)

        owner.setLastSentLength(cumulativeText.count)
        owner.sink()?.onLlmEvent(LlmEventData(
            sessionId: owner.agentId(),
            requestId: msgId,
            kind: "firstToken",
            textDelta: delta
        ))
    }

    func onSentenceDone(text: String) {
        guard let owner = owner else { return }
        owner.setLastSentLength(0)
        if let msgId = owner.getCurrentAssistantId() {
            owner.setCurrentAssistantId(nil)
            owner.sink()?.onLlmEvent(LlmEventData(
                sessionId: owner.agentId(),
                requestId: msgId,
                kind: "done",
                fullText: text
            ))
        } else {
            // No streaming bubble yet — emit a single-shot reply.
            let id = UUID().uuidString
            owner.sink()?.onLlmEvent(LlmEventData(
                sessionId: owner.agentId(),
                requestId: id,
                kind: "firstToken",
                textDelta: text
            ))
            owner.sink()?.onLlmEvent(LlmEventData(
                sessionId: owner.agentId(),
                requestId: id,
                kind: "done",
                fullText: text
            ))
        }
    }

    func onDisconnected() {
        guard let owner = owner else { return }
        owner.transitionTo(.idle)
        owner.sink()?.onConnectionStateChanged(
            sessionId: owner.agentId(),
            state: "disconnected",
            errorMessage: nil
        )
    }

    func onError(code: String, message: String) {
        guard let owner = owner else { return }
        owner.transitionTo(.error)
        owner.sink()?.onConnectionStateChanged(
            sessionId: owner.agentId(),
            state: "error",
            errorMessage: message
        )
        owner.sink()?.onError(
            sessionId: owner.agentId(),
            errorCode: code,
            message: message,
            requestId: nil
        )
        owner.sink()?.onAgentReady(
            sessionId: owner.agentId(),
            ready: false,
            errorCode: code,
            errorMessage: message
        )
    }

    func onSpeechStart() {
        guard let owner = owner else { return }
        if let prevId = owner.getCurrentAssistantId() {
            owner.setLastSentLength(0)
            owner.sink()?.onLlmEvent(LlmEventData(
                sessionId: owner.agentId(),
                requestId: prevId,
                kind: "done"
            ))
            owner.setCurrentAssistantId(nil)
        }
        owner.sink()?.onSttEvent(SttEventData(
            sessionId: owner.agentId(),
            requestId: "",
            kind: "vadSpeechStart"
        ))
    }

    func onStateChanged(state: String) {
        guard let owner = owner else { return }
        owner.sink()?.onStateChanged(
            sessionId: owner.agentId(),
            state: state,
            requestId: nil
        )
    }
}
